ThirdPartyAdapters/ironsource/ironsource/src/main/java/com/google/ads/mediation/ironsource/IronSourceMediationAdapter.java

```diff
@@ -174,248 +124,105 @@ public void initialize(@NonNull Context context,
+      String message =
+          String.format(
+              "Multiple '%s' entries found: %s. Using app key '%s' to initialize the IronSource SDK.",
               KEY_APP_KEY, appKeys, appKey);
       Log.w(TAG, message);
     }
 
-    IronSourceManager.getInstance().initIronSourceSDK(context, appKey,
-        new IronSourceManager.InitializationCallback() {
-          @Override
-          public void onInitializeSuccess() {
-            initializationCompleteCallback.onInitializationSucceeded();
-          }
+    IronSource.setMediationType(MEDIATION_NAME + ADAPTER_VERSION_NAME);
+    Log.d(TAG, "Initializing IronSource SDK with app key: " + appKey);
+    IronSource.initISDemandOnly(
+        context,
+        appKey,
+        IronSource.AD_UNIT.INTERSTITIAL,
+        IronSource.AD_UNIT.REWARDED_VIDEO,
+        IronSource.AD_UNIT.BANNER);
+    isInitialized.set(true);
+    initializationCompleteCallback.onInitializationSucceeded();
 
-          @Override
-          public void onInitializeError(@NonNull AdError initializationError) {
-            initializationCompleteCallback.onInitializationFailed(initializationError.getMessage());
-          }
-        });
+    IronSource.setISDemandOnlyInterstitialListener(
+        IronSourceInterstitialAd.getIronSourceInterstitialListener());
+    IronSource.setISDemandOnlyRewardedVideoListener(
+        IronSourceRewardedAd.getIronSourceRewardedListener());
   }
 
   @Override
   public void loadRewardedAd(
       @NonNull MediationRewardedAdConfiguration mediationRewardedAdConfiguration,
-      @NonNull final MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
-          mediationAdLoadCallback) {
-
-    Bundle serverParameters = mediationRewardedAdConfiguration.getServerParameters();
-    Context context = mediationRewardedAdConfiguration.getContext();
-    String appKey = serverParameters.getString(KEY_APP_KEY);
-    this.instanceID = serverParameters.getString(KEY_INSTANCE_ID, DEFAULT_INSTANCE_ID);
-
-    IronSourceManager.getInstance().initIronSourceSDK(context, appKey,
-        new InitializationCallback() {
-          @Override
-          public void onInitializeSuccess() {
-            IronSourceMediationAdapter.this.mediationAdLoadCallback = mediationAdLoadCallback;
-            Log.d(TAG,
-                String.format("Loading IronSource rewarded ad with instance ID: %s", instanceID));
-            IronSourceManager.getInstance()
-                .loadRewardedVideo(context, instanceID, IronSourceMediationAdapter.this);
-          }
-
-          @Override
-          public void onInitializeError(@NonNull AdError initializationError) {
-            Log.e(TAG, initializationError.getMessage());
-            mediationAdLoadCallback.onFailure(initializationError);
-          }
-        });
+      @NonNull
+          final MediationAdLoadCallback<MediationRewardedAd, MediationRewardedAdCallback>
+              mediationAdLoadCallback) {
+    if (isInitialized.get()) {
```

Comment:
We are checking `isInitialized()` before loading ads. But if it is not initialized, then we do nothing, which I believe will cause an adapter timeout with 0 feedback to the publisher.

We should attempt to initialize the IronSource SDK, if not already initialized, at load time, on each respective class' `loadAd()` method.

Reply:
We have added a logic to return load fail in case of loading when the SDK is not initialized.

Reply:
Hey @joshh-devrel (Ophir here, from Alon's team), thanks for the question. We find that having a possible call to our SDK initialization on every ad load can lead to some issues. We currently prefer keeping the current implementation where we initialize our SDK in a single place, and an init failure on our side means we will not take part in the current session. The same applies to all ad units.

Reply:
Thanks @Ophir-IS. A follow-up, since we did `isInitialized.set(true);` above, will this statement `if (isInitialized.get())` always be true?

Reply:
Assuming `initialize()` is always called before `loadRewardedAd()` (or any other ad unit load function), then yes this is correct.

Reply:
If we are keeping this validation, may I suggest removing the if-else statement and simplify it at the start of each load method for easier readability like so:
```suggestion
    if (!isInitialized.get()) {
      AdError adError = ... ;
      mediationAdLoadCallback.onFailure(adError);
      return;
    }
```

Reply:
Done